### PR TITLE
Lenet mods + NiN(new addition) + train.py improvement

### PR DESCRIPTION
**Pull request by saswatpp**
1.Lenet is tested on the new train_ch5() function
2.train_ch5() has following chanages
	a. Loss and accuracy calculated on torch tensors on the device so it improves speed
	b. y is not typecasted to float for comparision. Instead as argmax() returns a long tensor, y is compared as it is.
	c. Accuracy calculation is done inside torch.no_grad() block . No need for detach()
3. Evaluate Accuracy has the following changes
	a. X and y are directly collected from test_iter rather than using second for loop and zipping features and labels( like in old one). It utilizes vectorized implementation rather than usage of loops
	b. acc_sum is now a tensor in the specified device(improved speed)
	c. torch.no_grad() removes the requirement of detach()
4. Network in Network Model was added (tested in Google Colab GPU with better accuracy than book even with same paramters(lr,batch,init)

### CODE REVIEW

File: d2l/train.py [hunk 1]
```diff
@@ -9,19 +9,18 @@
 
 __all__ = ['evaluate_accuracy', 'train_ch3', 'train_ch5']
 
-def evaluate_accuracy(data_iter, net, device=torch.device('cpu')):
+def evaluate_accuracy(data_iter, net,device=torch.device('cpu')):
```
Comment (AnirudhDagar) [inline d2l/train.py]: keep the space between each arg, like before
```suggestion
def evaluate_accuracy(data_iter, net, device=torch.device('cpu')):
```

File: d2l/train.py [hunk 2]
```diff
@@ -46,27 +45,32 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
               % (epoch + 1, train_l_sum / n, train_acc_sum / n, test_acc))
 
 
-def train_ch5(net, train_iter, test_iter, criterion, num_epochs, batch_size, device, lr=None):
-    net = net.to(device)
+def train_ch5(net, train_iter, test_iter,criterion, num_epochs, batch_size, device,lr=None):
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion
def train_ch5(net, train_iter, test_iter, criterion, num_epochs, batch_size, device, lr=None):
```

File: d2l/train.py [hunk 3]
```diff
@@ -9,19 +9,18 @@
...
-            y = y.type(torch.float32)
-            X, y = X.to(device), y.to(device)
-            acc_sum += torch.sum((torch.argmax(net(X), dim=1).type(torch.float32) == y).detach()).float()
-            n += 1
-    return acc_sum.item() / n
+    acc_sum,n = torch.tensor([0],dtype=torch.float32,device=device),0
+    for X,y in data_iter:
+        # If device is the GPU, copy the data to the GPU.
+        X,y = X.to(device),y.to(device)
+        net.eval()
```
Comment (AnirudhDagar) [inline d2l/train.py]: let the eval mode be at the top(just after the func starts) with the comment. It is more readable there.
```
net.eval()  # Switch to evaluation mode for Dropout, BatchNorm etc layers.
```

File: d2l/train.py [hunk 4]
```diff
@@ -9,19 +9,18 @@
...
-        features, labels = batch
-        for X, y in zip(features, labels):
-            # If device is the GPU, copy the data to the GPU.
-            y = y.type(torch.float32)
-            X, y = X.to(device), y.to(device)
-            acc_sum += torch.sum((torch.argmax(net(X), dim=1).type(torch.float32) == y).detach()).float()
-            n += 1
-    return acc_sum.item() / n
+    acc_sum,n = torch.tensor([0],dtype=torch.float32,device=device),0
+    for X,y in data_iter:
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion
    for X, y in data_iter:
```

File: d2l/train.py [hunk 5]
```diff
@@ -9,19 +9,18 @@
...
+        # If device is the GPU, copy the data to the GPU.
+        X,y = X.to(device),y.to(device)
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion
        X, y = X.to(device), y.to(device)
```

File: d2l/train.py [hunk 6]
```diff
@@ -46,27 +45,32 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
...
     optimizer = optim.SGD(net.parameters(), lr=lr)
     for epoch in range(num_epochs):
-        net.train() # Switch to training mode
-        train_l_sum, train_acc_sum, n, start = 0.0, 0.0, 0, time.time()
+        train_l_sum = torch.tensor([0.0],dtype=torch.float32,device=device)
+        train_acc_sum = torch.tensor([0.0],dtype=torch.float32,device=device)
+        n, start = 0, time.time()
         for X, y in train_iter:
-            X, y = X.to(device), y.to(device)
+            net.train()
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion
            net.train() # Switch to training mode
```

File: d2l/train.py [hunk 7]
```diff
@@ -46,27 +45,32 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
...
-        print('epoch %d, loss %.4f, train acc %.3f, test acc %.3f, time %.1f sec'
-              % (epoch + 1, train_l_sum / n, train_acc_sum / n, test_acc, time.time() - start))
+            
+            with torch.no_grad():
+                y = y.long()
+                train_l_sum += loss.float()
+                train_acc_sum += (torch.sum((torch.argmax(y_hat, dim=1) == y))).float()
+                n += y.shape[0]
+        test_acc = evaluate_accuracy(test_iter, net,device)
+        print('epoch %d, loss %.4f, train acc %.3f, test acc %.3f, '
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion
         print('epoch %d, loss %.4f, train acc %.3f, test acc %.3f, time %.1f sec'
```

File: d2l/train.py [hunk 8]
```diff
@@ -46,27 +45,32 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
...
+              'time %.1f sec'
```
Comment (AnirudhDagar) [inline d2l/train.py]:
```suggestion

```

File: d2l/train.py [hunk 9]
```diff
@@ -47,26 +46,30 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
...
     optimizer = optim.SGD(net.parameters(), lr=lr)
     for epoch in range(num_epochs):
-        net.train() # Switch to training mode
-        train_l_sum, train_acc_sum, n, start = 0.0, 0.0, 0, time.time()
+        train_l_sum = torch.tensor([0.0],dtype=torch.float32,device=device)
+        train_acc_sum = torch.tensor([0.0],dtype=torch.float32,device=device)
+        n, start = 0, time.time()
         for X, y in train_iter:
-            X, y = X.to(device), y.to(device)
+            net.train() # Switch to training mode
```
Comment (AnirudhDagar) [inline d2l/train.py]: Again like net.eval please take it to the original position. We don't want to call this in each loop for a single iterator. Take it below the line 
```
for epoch in range(num_epochs):
    net.train() # Switch to training mode

```

File: d2l/train.py (hunk 3 above)
Comment (saswatpp) [reply d2l/train.py]: The following error is resolved.

File: d2l/train.py (hunk 9 above)
Comment (saswatpp) [reply d2l/train.py]: The following Error is resolved.

File: d2l/train.py [hunk 10]
```diff
@@ -47,26 +46,28 @@ def train_ch3(net, train_iter, test_iter, criterion, num_epochs, batch_size, lr=
...
-
-            y = y.type(torch.float32)
-            y.to(device)
-            train_l_sum += loss.item()
-            train_acc_sum += torch.sum((torch.argmax(y_hat, dim=1).type(torch.float32) == y).detach()).float()
-            n += list(y.size())[0]
-        test_acc = evaluate_accuracy(test_iter, net, device)
-        print('epoch %d, loss %.4f, train acc %.3f, test acc %.3f, time %.1f sec'
-              % (epoch + 1, train_l_sum / n, train_acc_sum / n, test_acc, time.time() - start))
+
```
Comment (AnirudhDagar) [inline d2l/train.py]: no need to leave a blank line here.

Comment (saswatpp) [reply d2l/train.py]: Issue resolved
